### ISSUE TITLE
Struggling with ubsan suppressions

### ISSUE THREAD

**Issue by mdavidsaver** [llvm/llvm-project#72060]
Please:

- Include exact suppression name in sanitizer error messages
- Relax the suppressions parser to warn on unknown suppression name.
- Improve error messages from the suppressions parser.  (at least a line number)

tldr...

I would like to enable `-fsanitize=undefined` to some of my existing projects as a continuous integration test.  Being legacy C/C++ code, some UB is flagged.  To my mind, a natural starting point is to add suppressions for the existing cases so that I can enable this in CI immediately to prevent any new UB from being introduced.  Should be simply right?  Well, unfortunately not for me.  Perhaps I was not looking in the [correct](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) [places](https://clang.llvm.org/docs/SanitizerSpecialCaseList.html).

Firstly, the parser for `UBSAN_OPTIONS=suppressions=...` files treats unknown suppression names are a hard error.  Also, any such error results in a frustratingly brief error message.

```
UndefinedBehaviorSanitizer: failed to parse suppressions
```

No hint about why.  Not even a line number!

Next there is the question of suppression names.  Errors are printed like:

```
ubsan-check.c:4:29: runtime error: nan is outside the range of representable values of type 'long'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ubsan-check.c:4:29 in 
ubsan-check.c:8:19: runtime error: shift exponent 32 is too large for 32-bit type 'int'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ubsan-check.c:8:19 in 
```

I my particular case I would like to suppress the "nan" cast, but not the (over)shift.  How to do this?

I naively started with:

```
undefined-behavior:
```

nope, `failed to parse suppressions`.

then based on the documentation I tried:

```
undefined:
```

which parses, but does not suppress either.

To cut short a long story.  I eventually found where the suppression names appear in the llvm source tree.  I downloaded a local copy of `compiler-rt/lib/ubsan/ubsan_checks.inc` and run it through the following to generate a file to suppress every ubsan check.  Then by process of elimination I found that `float-cast-overflow:` and `shift-exponent:` were the relevant suppression names.

```c
#include <stdio.h>
#define UBSAN_CHECK(X, Y, Z) printf("%s:\n", Z);
int main() {
#include "ubsan_checks.inc"
    return 0;
}
```

So now I see a realistic path to enabling sanitizers in my CI builds.  However, the strictness of the suppressions file parser gives me some concern.  Can I expect suppression names to be the same with older or newer version of llvm?

**Comment by mdavidsaver** [comment]
Based on experience with `-march=?`, I thought that `-fsanitize=?` might print a list of suppression names.  But no such luck.

**Comment by mdavidsaver** [comment]
I am testing on Debian 12 with packaged clang `1:14.0-55.7~deb12u1`.

```
$ clang --version
Debian clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by firewave** [comment]
This is a rather long-standing issue which I ran into several years ago but never reported upstream (but maybe I did and just forgot - at least I could not longer find any related tickets).

The main issue with all sanitizers is the incomplete documentation (let's be honest - any proper documentation at all) of the `*SAN_OPTIONS` environment variables (they do exist within the source though: https://github.com/gcc-mirror/gcc/blob/master/libsanitizer/ubsan/ubsan_flags.inc). Some of it is still living in the original source of the sanitizer documentation: https://github.com/google/sanitizers/wiki.

Regarding the suppression of UBSAN types you need to specify `report_error_type=1` via `UBSAN_OPTIONS`. That will change `undefined` to the proper type for a suppression.

Unfortunately some checks generate types which are unknown and thus cannot be suppressed. I cannot recall an example of those but I assume some simple unit tests should turn those up. It is possible though that this issue has been resolved since I ran into it the last times which was several years ago.

**Comment by schenker** [comment]
It also took me quite some time to figure out how to use the suppresion files. I created a PR to improve the error message a bit: https://github.com/llvm/llvm-project/pull/72066.

**Comment by mdavidsaver** [comment]
> ... you need to specify `report_error_type=1` via `UBSAN_OPTIONS` ...

Wow, very good to know.  This might have saved me an hour or so.  The name printed is still eg. the internal `invalid-shift-exponent` instead of the external `shift-exponent`.  Still, `invalid-shift-exponent` only appears once it the llvm source tree, so I would have found it sooner.

Since the (over)shift also produces a compile time warning, maybe useful to contrast the two messages.  

```
ubsan-shift.c:3:19: warning: shift count >= width of type [-Wshift-count-overflow]
```

manages to convey more information to me than the two ubsan lines combined, with and without `report_error_type=1`.  I especially value like the `[-Wshift-count-overflow]`, as experience of the conventions of clang/gcc would suggest `-Wno-shift-count-overflow`.

```
ubsan-shift.c:3:19: runtime error: shift exponent 32 is too large for 32-bit type 'int'
SUMMARY: UndefinedBehaviorSanitizer: invalid-shift-exponent ubsan-shift.c:3:19 in 
```

imo. there is some redundancy here.  A more concise expression might be:

```
ubsan-shift.c:3:19: error: shift exponent 32 >= width of 'int' [-fsanitize=shift-exponent]
```

**Comment by dwblaikie** [comment]
Perhaps you folks could propose/write some docs? (at least with whatever understanding you currently have - would be a start and something the experts who know how it works can weigh in on, provide corrections to and then get it committed?)

**Comment by mariadb-RoelVandePaar** [comment]
Indeed, the syntax of UBSAN filtering is not the most clear/obvious. For an example of currently working filtering with some extra info, see: https://github.com/mariadb-corporation/mariadb-qa/blob/master/UBSAN.filter
